File: clients/mobile/lib/utils/permission_util.dart
class PermissionGroup {
  PermissionGroup({required this.module, required this.label, required this.permissions});

  final String module;
  final String label;
  final List<String> permissions;
}

String permissionModule(String permission) {
  if (permission == '*.*' || permission.endsWith('.*')) {
    return 'platform';
  }
  final dot = permission.indexOf('.');
  return dot > 0 ? permission.substring(0, dot) : 'platform';
}

List<PermissionGroup> groupPermissions(List<String> permissions) {
  final map = <String, List<String>>{};
  for (final permission in permissions) {
    final module = permissionModule(permission);
    map.putIfAbsent(module, () => []).add(permission);
  }
  final entries = map.entries.toList()..sort((a, b) => a.key.compareTo(b.key));
  return entries
      .map(
        (entry) => PermissionGroup(
          module: entry.key,
          label: _formatModuleLabel(entry.key),
          permissions: [...entry.value]..sort(),
        ),
      )
      .toList();
}

bool hasSelectedPermission(List<String> selected, String permission) {
  if (selected.contains(permission) || selected.contains('*.*')) {
    return true;
  }
  final module = permissionModule(permission);
  if (selected.contains('$module.*')) {
    return true;
  }
  if (selected.contains('admin.*') && module == 'admin') {
    return true;
  }
  for (final entry in selected) {
    if (entry.endsWith('.*')) {
      final prefix = entry.substring(0, entry.length - 1);
      if (permission.startsWith(prefix)) {
        return true;
      }
    }
  }
  return false;
}

List<String> togglePermission(List<String> selected, String permission, bool enabled) {
  final next = selected.where((entry) => !entry.endsWith('.*') || entry == permission).toSet();
  if (enabled) {
    next.add(permission);
  } else {
    next.remove(permission);
  }
  return next.toList()..sort();
}

List<String> toggleWildcard(List<String> selected, String wildcard, bool enabled) {
  final module = wildcard.replaceAll('.*', '');
  final withoutModule = selected
      .where((entry) => entry != wildcard && permissionModule(entry) != module)
      .toList();
  if (enabled) {
    return [...withoutModule, wildcard]..sort();
  }
  return withoutModule..sort();
}

String formatRoleSummary(List<Map<String, dynamic>> roles) {
  if (roles.isEmpty) {
    return '—';
  }
  return roles.map((role) => '${role['name'] ?? role['code']}').join(', ');
}

String _formatModuleLabel(String code) {
  if (code == 'platform') {
    return 'Platform';
  }
  return code
      .split('_')
      .map((part) => part.isEmpty ? part : '${part[0].toUpperCase()}${part.substring(1)}')
      .join(' ');
}

File: clients/web/src/app/shared/utils/permission.util.ts
export interface PermissionGroup {
  module: string;
  label: string;
  permissions: string[];
}

const MODULE_LABELS: Record<string, string> = {
  admin: 'Administration',
  customer: 'Customer',
  product: 'Product',
  inventory: 'Inventory',
  crm: 'CRM',
  platform: 'Platform',
};

export function formatModuleLabel(code: string): string {
  return (
    MODULE_LABELS[code] ??
    code
      .split('_')
      .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
      .join(' ')
  );
}

export function permissionModule(permission: string): string {
  if (permission === '*.*' || permission.endsWith('.*')) {
    return 'platform';
  }
  const dot = permission.indexOf('.');
  return dot > 0 ? permission.slice(0, dot) : 'platform';
}

export function groupPermissions(permissions: string[]): PermissionGroup[] {
  const map = new Map<string, string[]>();
  for (const permission of permissions) {
    const module = permissionModule(permission);
    const list = map.get(module) ?? [];
    list.push(permission);
    map.set(module, list);
  }
  return Array.from(map.entries())
    .sort(([left], [right]) => left.localeCompare(right))
    .map(([module, items]) => ({
      module,
      label: formatModuleLabel(module),
      permissions: [...items].sort(),
    }));
}

export function hasSelectedPermission(selected: string[], permission: string): boolean {
  if (selected.includes(permission) || selected.includes('*.*')) {
    return true;
  }
  const module = permissionModule(permission);
  if (selected.includes(`${module}.*`)) {
    return true;
  }
  if (selected.includes('admin.*') && module === 'admin') {
    return true;
  }
  for (const entry of selected) {
    if (entry.endsWith('.*')) {
      const prefix = entry.slice(0, -1);
      if (permission.startsWith(prefix)) {
        return true;
      }
    }
  }
  return false;
}

export function togglePermission(selected: string[], permission: string, enabled: boolean): string[] {
  const next = new Set(selected.filter((entry) => !entry.endsWith('.*') || entry === permission));
  if (enabled) {
    next.add(permission);
  } else {
    next.delete(permission);
  }
  return Array.from(next).sort();
}

export function toggleWildcard(selected: string[], wildcard: string, enabled: boolean): string[] {
  const module = wildcard.replace('.*', '');
  const withoutModule = selected.filter(
    (entry) => entry !== wildcard && permissionModule(entry) !== module,
  );
  if (enabled) {
    return [...withoutModule, wildcard].sort();
  }
  return withoutModule.sort();
}

export function formatRoleSummary(roles: Array<{ code: string; name: string }>): string {
  if (roles.length === 0) {
    return '—';
  }
  return roles.map((role) => role.name || role.code).join(', ');
}

export function permissionGroupSummary(permissions: string[]): string {
  if (permissions.length === 0) {
    return '—';
  }
  return groupPermissions(permissions)
    .map((group) => `${group.label} (${group.permissions.length})`)
    .join(', ');
}
